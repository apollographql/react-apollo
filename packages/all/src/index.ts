// Common + hooks
export * from '@apollo/react-common';

// @apollo/react-components
export {
  Query,
  Mutation,
  Subscription,
  QueryComponentOptions,
  MutationComponentOptions,
  SubscriptionComponentOptions
} from '@apollo/react-components';

// @apollo/react-hoc
export {
  graphql,
  withQuery,
  withMutation,
  withSubscription,
  withApollo,
  QueryControls,
  DataValue,
  DataProps,
  MutateProps,
  ChildProps,
  ChildDataProps,
  ChildMutateProps,
  OptionProps,
  OperationOption,
  WithApolloClient
} from '@apollo/react-hoc';

// SSR
export {
  getMarkupFromTree,
  getDataFromTree,
  renderToStringWithData
} from '@apollo/react-ssr';
