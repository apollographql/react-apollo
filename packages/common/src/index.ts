export {
  ApolloContextValue,
  ApolloContextOptions,
  getApolloContext,
  resetApolloContext
} from './context/ApolloContext';
export { ApolloProvider } from './context/ApolloProvider';
export { ApolloConsumer } from './context/ApolloConsumer';

export {
  parser,
  operationName,
  DocumentType,
  IDocumentDefinition
} from './parser/parser';

export * from './types/types';
