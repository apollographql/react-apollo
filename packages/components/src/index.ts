export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export { Query } from './Query';
export { Mutation } from './Mutation';
export { Subscription } from './Subscription';

export * from './types';
