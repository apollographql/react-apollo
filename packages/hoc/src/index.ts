export { graphql } from './graphql';

export { withQuery } from './query-hoc';
export { withMutation } from './mutation-hoc';
export { withSubscription } from './subscription-hoc';
export { withApollo } from './withApollo';

export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export * from './types';
