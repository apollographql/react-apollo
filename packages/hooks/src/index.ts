export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export { useQuery } from './useQuery';
export { useLazyQuery } from './useLazyQuery';
export { useMutation } from './useMutation';
export { useSubscription } from './useSubscription';
export { useApolloClient } from './useApolloClient';

export { RenderPromises } from './ssr/RenderPromises';

export * from './types';
