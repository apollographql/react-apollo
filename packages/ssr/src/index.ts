export { getMarkupFromTree, getDataFromTree } from './getDataFromTree';
export { renderToStringWithData } from './renderToStringWithData';
export * from './types';
