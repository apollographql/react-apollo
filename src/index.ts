export * from './browser';
export * from './server';
