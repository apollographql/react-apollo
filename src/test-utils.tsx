import * as React from 'react';
import ApolloClient from 'apollo-client';

import {
  NetworkInterface,
  Request,
} from 'apollo-client/networkInterface';

import {
  GraphQLResult,
  Document,
} from 'graphql';

import { print } from 'graphql-tag/printer';


import ApolloProvider from './ApolloProvider';

export class MockedProvider extends React.Component<any, any> {
  private client: any;

  constructor(props, context) {
    super(props, context);

    const networkInterface = mockNetworkInterface.apply(null, this.props.mocks);
    this.client = new ApolloClient({ networkInterface });
  }

  render() {
    return (
      <ApolloProvider client={this.client}>
        {this.props.children}
      </ApolloProvider>
    );
  }
}

// Pass in multiple mocked responses, so that you can test flows that end up
// making multiple queries to the server
export function mockNetworkInterface(
  ...mockedResponses: MockedResponse[]
): NetworkInterface {
  return new MockNetworkInterface(...mockedResponses);
}

export interface ParsedRequest {
  variables?: Object;
  query?: Document;
  debugName?: string;
}

export interface MockedResponse {
  request: ParsedRequest;
  result?: GraphQLResult;
  error?: Error;
  delay?: number;
  newData?: () => any;
}

export class MockNetworkInterface implements NetworkInterface {
  private mockedResponsesByKey: { [key: string]: MockedResponse[] } = {};

  constructor(...mockedResponses: MockedResponse[]) {
    mockedResponses.forEach((mockedResponse) => {
      this.addMockedReponse(mockedResponse);
    });
  }

  public addMockedReponse(mockedResponse: MockedResponse) {
    const key = requestToKey(mockedResponse.request);
    let mockedResponses = this.mockedResponsesByKey[key];
    if (!mockedResponses) {
      mockedResponses = [];
      this.mockedResponsesByKey[key] = mockedResponses;
    }
    mockedResponses.push(mockedResponse);
  }

  public query(request: Request) {
    return new Promise((resolve, reject) => {
      const parsedRequest: ParsedRequest = {
        query: request.query,
        variables: request.variables,
        debugName: request.debugName,
      };

      const key = requestToKey(parsedRequest);

      if (!this.mockedResponsesByKey[key]) {
        throw new Error('No more mocked responses for the query: ' + print(request.query));
      }

      const original = [...this.mockedResponsesByKey[key]];
      const { result, error, delay, newData } = this.mockedResponsesByKey[key].shift() || {} as any;

      if (newData) {
        original[0].result = newData();
        this.mockedResponsesByKey[key].push(original[0]);
      }

      if (!result && !error) {
        throw new Error(`Mocked response should contain either result or error: ${key}`);
      }

      setTimeout(() => {
        if (error) return reject(error);
        return resolve(result);
      }, delay ? delay : 1);
    });
  }
}

function requestToKey(request: ParsedRequest): string {
  const queryString = request.query && print(request.query);
  return JSON.stringify({
    variables: request.variables,
    debugName: request.debugName,
    query: queryString,
  });
}
