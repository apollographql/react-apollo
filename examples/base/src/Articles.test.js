import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from "../../../lib/src/test-utils";
import { print } from 'graphql-tag/printer';

import ArticlesWithData, { ARTICLE_QUERY, Articles, withArticles } from "./Articles";

const mockedData = {
  content: [
    {
      title: "Mollis Bibendum Vulputate Commodo",
      meta: {
        summary: "Maecenas faucibus mollis interdum. Nulla vitae elit libero, a pharetra augue. Sed posuere consectetur est at lobortis."
      },
    },
    {
      title: "Ullamcorper Fusce Egestas",
      meta: {
        summary: "Maecenas faucibus mollis interdum. Donec sed odio dui. Donec id elit non mi porta gravida at eget metus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Curabitur blandit tempus porttitor."
      },
    },
    {
      title: "Tristique Sollicitudin Tellus Bibendum Vehicula",
      meta: {
        summary: "Sed posuere consectetur est at lobortis. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Nulla vitae elit libero, a pharetra augue. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Nulla vitae elit libero, a pharetra augue. Vestibulum id ligula porta felis euismod semper."
      },
    },
  ],
};

describe('default export', () => {
  it('renders without crashing', () => {
    const output = renderer.create(
      <MockedProvider mocks={[{ request: { query: ARTICLE_QUERY }, result: { data: mockedData } }]}>
        <ArticlesWithData />
      </MockedProvider>
    )
    expect(output.toJSON()).toMatchSnapshot();
  });
});

describe('Article enhancer', () => {
  it('renders with loading first', (done) => {
    class Container extends React.Component {
      componentWillMount() {
        expect(this.props.data.loading).toBe(true);
        expect(this.props.data.content).toBeFalsy();
        done();
      }
      render() {
        return null;
      }
    };
    const ContainerWithData = withArticles(Container);
    const output = renderer.create(
      <MockedProvider mocks={[{ request: { query: ARTICLE_QUERY }, result: { data: mockedData } }]}>
        <ContainerWithData />
      </MockedProvider>
    );
  });

   it('renders data without crashing', (done) => {
    class Container extends React.Component {
      componentWillReceiveProps(props) {
        expect(props.data.loading).toBe(false);
        expect(props.data.content).toEqual(mockedData.content);
        done();
      }
      render() {
        return null;
      }
    };
    const ContainerWithData = withArticles(Container);
    const output = renderer.create(
      <MockedProvider mocks={[{ request: { query: ARTICLE_QUERY }, result: { data: mockedData } }]}>
        <ContainerWithData />
      </MockedProvider>
    );
  });

});

describe('Article query', () => {
  it('should match expected structure', () => {
    expect(ARTICLE_QUERY).toMatchSnapshot();
  });

  it('should match expected shape', () => {
    expect(print(ARTICLE_QUERY)).toMatchSnapshot();
  });
});

describe('Articles Component', () => {
  it('should render a loading state without data', () => {
    const output = renderer.create(<Articles data={{ loading: true }} />)
    expect(output.toJSON()).toMatchSnapshot();
  });

  it('should render title and summary in order', () => {
    const output = renderer.create(
      <Articles data={{ loading: false, content: mockedData.content }} />
    );
    expect(output.toJSON()).toMatchSnapshot();
  });
});


