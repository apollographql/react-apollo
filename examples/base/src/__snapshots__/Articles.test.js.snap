exports[`Article query should match expected shape 1`] = `
"query GetArticles {
  content(channel: \"articles\", limit: 3) {
    title
    meta {
      summary
    }
  }
}
"
`;

exports[`Article query should match expected structure 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "GetArticles"
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "channel"
                },
                "value": Object {
                  "kind": "StringValue",
                  "value": "articles"
                }
              },
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "limit"
                },
                "value": Object {
                  "kind": "IntValue",
                  "value": "3"
                }
              }
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "content"
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "title"
                  },
                  "selectionSet": null
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "meta"
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "summary"
                        },
                        "selectionSet": null
                      }
                    ]
                  }
                }
              ]
            }
          }
        ]
      },
      "variableDefinitions": Array []
    }
  ],
  "kind": "Document"
}
`;

exports[`Articles Component should render a loading state without data 1`] = `
<div>
  Loading
</div>
`;

exports[`Articles Component should render title and summary in order 1`] = `
<div>
  <div>
    <h3>
      Mollis Bibendum Vulputate Commodo
    </h3>
    <p>
      Maecenas faucibus mollis interdum. Nulla vitae elit libero, a pharetra augue. Sed posuere consectetur est at lobortis.
    </p>
  </div>
  <div>
    <h3>
      Ullamcorper Fusce Egestas
    </h3>
    <p>
      Maecenas faucibus mollis interdum. Donec sed odio dui. Donec id elit non mi porta gravida at eget metus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Curabitur blandit tempus porttitor.
    </p>
  </div>
  <div>
    <h3>
      Tristique Sollicitudin Tellus Bibendum Vehicula
    </h3>
    <p>
      Sed posuere consectetur est at lobortis. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Nulla vitae elit libero, a pharetra augue. Vivamus sagittis lacus vel augue laoreet rutrum faucibus dolor auctor. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Nulla vitae elit libero, a pharetra augue. Vestibulum id ligula porta felis euismod semper.
    </p>
  </div>
</div>
`;

exports[`default export renders without crashing 1`] = `
<div>
  Loading
</div>
`;
